Sources/Datadog/DatadogConfiguration.swift

```diff
@@ -62,10 +63,19 @@ extension Datadog {
                 return self
             }
 
+            internal func set(dateProvider: DateProvider) -> Builder {
+                self.dateProvider = dateProvider
+                return self
+            }
+
             /// Builds `Datadog.Configuration` object.
             public func build() -> Configuration {
                 return Configuration(
-                    logsUploadURL: try? DataUploadURL(endpointURL: logsEndpoint.url, clientToken: clientToken)
+                    logsUploadURLProvider: try? UploadURLProvider(
```

Comment:
Hmm, IMO we have two problems in `DatadogConfiguration`:
* 1/ The `set(dateProvider: DateProvider) -> Builder` is there only for unit tests;
* 2/ We don't leverage `throwable` initializer of `UploadURLProvider`, so the `ProgrammerErrors` thrown from there are not printed to the user.

IMO we can fix both by simply changing from:
```swift
struct DatadogConfiguration {
   internal let logsUploadURLProvider: UploadURLProvider?
   // ...
}
```
to:
```swift
struct DatadogConfiguration {
   internal let endpointURL: URL
   internal let clientToken: String
   // ...
}
```
and delegating `UploadURLProvider.init(...)` to `Datadog`.

`Datadog` consumes `DatadogConfiguration` in:
```swift
// Datadog.swift
private static func initializeOrThrow(appContext: AppContext, configuration: Configuration) throws 
```
so we might `rethrow` the error from `UploadURLProvider` and it will be printed for the user.

This will solve both 1/ and 2/. Also, this will print `ProgrammerError` to the user if `UploadURLProvider` cannot be constructed, which we can unit test in `DatadogTests`.

Reply:
super idea 🙌 
i didn't like what i did neither 😕 
i'm changing ♻️